receiver/loadgenreceiver/logs.go

```diff
@@ -0,0 +1,132 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package loadgenreceiver // import "github.com/elastic/opentelemetry-collector-components/receiver/loadgenreceiver"
+
+import (
+	"bufio"
+	"bytes"
+	"context"
+	_ "embed"
+	"os"
+	"time"
+
+	"github.com/elastic/opentelemetry-collector-components/receiver/loadgenreceiver/internal"
+	"go.opentelemetry.io/collector/component"
+	"go.opentelemetry.io/collector/consumer"
+	"go.opentelemetry.io/collector/pdata/pcommon"
+	"go.opentelemetry.io/collector/pdata/plog"
+	"go.opentelemetry.io/collector/receiver"
+	"go.uber.org/zap"
+)
+
+//go:embed testdata/logs.jsonl
+var demoLogs []byte
+
+type logsGenerator struct {
+	cfg    *Config
+	logger *zap.Logger
+
+	samples  internal.LoopingList[plog.Logs]
+	consumer consumer.Logs
+
+	cancelFn context.CancelFunc
+}
+
+func createLogsReceiver(
+	ctx context.Context,
+	set receiver.Settings,
+	config component.Config,
+	consumer consumer.Logs,
+) (receiver.Logs, error) {
+	genConfig := config.(*Config)
+
+	parser := plog.JSONUnmarshaler{}
+	var err error
+	sampleLogs := demoLogs
+
+	if genConfig.Logs.JsonlFile != "" {
+		sampleLogs, err = os.ReadFile(string(genConfig.Logs.JsonlFile))
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	var samples []plog.Logs
+	scanner := bufio.NewScanner(bytes.NewReader(sampleLogs))
+	for scanner.Scan() {
+		logBytes := scanner.Bytes()
+		lineLogs, err := parser.UnmarshalLogs(logBytes)
+		if err != nil {
+			return nil, err
+		}
+		samples = append(samples, lineLogs)
+	}
+
+	return &logsGenerator{
+		cfg:      genConfig,
+		logger:   set.Logger,
+		consumer: consumer,
+		samples:  internal.NewLoopingList(samples),
+	}, nil
+}
+
+func (ar *logsGenerator) Start(ctx context.Context, _ component.Host) error {
+	startCtx, cancelFn := context.WithCancel(ctx)
+	ar.cancelFn = cancelFn
+
+	go func() {
+		for {
+			select {
+			case <-startCtx.Done():
+				return
+			default:
+				if err := ar.consumer.ConsumeLogs(startCtx, ar.nextLogs()); err != nil {
+					ar.logger.Error(err.Error())
```

Comment:
If I add a rate limiter, this will print errors non-stop, and it is very hard to read the rest of the logs from my collector container. @carsonip How did you work around this? I have tried the filter processor, but from my understanding this filter only works to filter the actual logs from the loadgenreceiver, not the logs from the logger.

Reply:
> If I add a rate limiter, this will print errors non-stop, and it is very hard to read the rest of the logs from my collector container. @carsonip How did you work around this? I have tried the filter processor, but from my understanding this filter only works to filter the actual logs from the loadgenreceiver, not the logs from the logger.

ratelimitprocessor now returns an error whenever the receiver exceeds the rate limit. IMO we should make it configurable such that instead of returning an error immediately when it is throttled, it can optionally block until the events are sent. And the load generator should use this option, to reduce noise in logs.

Reply:
This is specifically the TODO stated in ratelimitprocessor: https://github.com/elastic/opentelemetry-collector-components/blob/76ebae8ff6a9e10094a7a417655c1ecdb040db87/processor/ratelimitprocessor/local.go#L60

Load generator should use delayed processing.

Reply:
draft PR: https://github.com/elastic/opentelemetry-collector-components/pull/293

Reply:
Fixed